route_auth.go

```diff
@@ -1,102 +1,242 @@
+			SetUsername(input.Username).
+			SetEmail(input.Email).
+			SetPassword(string(hashedPassword)).
+			SetVerifyToken(verifyToken).
+			Save(r.Context())
 		if err != nil {
 			return err
 		}
 
 		c, err := tx.Planet.Query().
-				Where(
-					planet.And(
-						planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
-						planet.Not(planet.HasOwner()),
-					),
-				).
-				Count(r.Context())
+			Where(
+				planet.And(
+					planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
+					planet.Not(planet.HasOwner()),
+				),
+			).
+			Count(r.Context())
 
 		if err != nil {
 			return err
 		}
 
-		ra := rand.New(rand.NewSource(42))
-
-		n := ra.Intn(c)
+		ra := rand.New(rand.NewSource(time.Now().UnixNano()))
+		rn := ra.Intn(c)
 
 		p, err := tx.Planet.Query().
-				Where(
-					planet.And(
-						planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
-						planet.Not(planet.HasOwner()),
-					),
-				).
-				Offset(n).
-				First(r.Context())
+			Where(
+				planet.And(
+					planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
+					planet.Not(planet.HasOwner()),
+				),
+			).
+			Offset(rn).
+			First(r.Context())
 
 		_, err = p.Update().
 			SetOwner(u).
 			Save(r.Context())
 
+		if err := sendSignupEmail(u); err != nil {
+			return err
+		}
+
 		return err
 	})
 	if err != nil {
 		serveError(w, r, newInternalServerError(fmt.Errorf("unable to create a user account: %v", err)))
 		return
 	}
 
-	flash(r, flashSuccess, "Your account has been created. You can log in now.")
-	http.Redirect(w, r, "/", 302)
+	// TODO: Move this to a separate page so the user isn't tempted to try to
+	//       log in even through they haven't activated their account yet.
+	flash(r, flashSuccess, "We have send you an email with a confirmation link")
+	http.Redirect(w, r, "/", 303)
+}
+
+var generateVerifyToken = func() (string, error) {
+	return generateRandomString(20)
+}
+
+func sendSignupEmail(u *ent.User) error {
+	tmpl, err := template.New("signup.html").ParseFiles("resources/emails/signup.html")
+	if err != nil {
+		return fmt.Errorf("could not parse signup email template: %w", err)
+	}
+	confirmUrl, err := absoluteUrl(fmt.Sprintf("confirm_email?email=%v&token=%v", url.QueryEscape(u.Email), u.VerifyToken))
+	if err != nil {
+		return fmt.Errorf("could not get confirmation url: %w", err)
+	}
+	var contents bytes.Buffer
+	err = tmpl.Execute(&contents, struct {
+		Username string
+		Url      string
+	}{
+		Username: u.Username,
+		Url:      confirmUrl,
+	})
+	if err != nil {
+		return fmt.Errorf("could not execute signup email template: %w", err)
+	}
+
+	return sendEmail(
+		u.Email,
+		u.Username,
+		"Welcome to Empire and Foundation",
+		&contents,
+	)
+}
+
+type confirmEmailRequest struct {
+	Email string `json:"email" name:"email" validate:"required"`
+	Token string `json:"token" name:"token" validate:"required"`
+}
+
+// GET /confirm_email
+// Activate an account and redirect to the login page
+func serveConfirmEmail(w http.ResponseWriter, r *http.Request) {
```

Comment:
Same, maybe we need another file for confirmation related code?

Reply:
It's still an authentication route. The `sendSignupEmail` was more of a helper to send mail.